public/pages/AnomalyCharts/utils/anomalyChartUtils.ts

```diff
@@ -295,6 +300,114 @@ export const getAnomaliesHeatmapData = (
+
+  const timeWindows = calculateTimeWindowsWithMaxDataPoints(
+    NUM_CELLS,
+    dateRange
+  );
+
+  let entitiesAnomalySummaries = [] as EntityAnomalySummaries[];
+
+  if (isEmpty(entitiesAnomalySummaryResult)) {
+    // put placeholder data so that heatmap won't look empty
+    for (let i = 0; i < displayTopNum; i++) {
+      // using blank string with different length as entity values instead of
+      // only 1 whitesapce for all entities, to avoid heatmap with single row
+      const blankStrValue = buildBlankStringWithLength(i);
+      entitiesAnomalySummaries.push({
+        entity: {
+          value: blankStrValue,
+        },
+      } as EntityAnomalySummaries);
+    }
+  } else {
+    entitiesAnomalySummaries = cloneDeep(entitiesAnomalySummaryResult);
+  }
+
+  entitiesAnomalySummaries.forEach((entityAnomalySummaries) => {
+    const maxAnomalyGradesForEntity = [] as number[];
+    const numAnomalyGradesForEntity = [] as number[];
+
+    const entityValue = get(entityAnomalySummaries, 'entity.value', '');
+    const anomaliesSummary = get(
+      entityAnomalySummaries,
+      'anomalySummaries',
+      []
+    ) as EntityAnomalySummary[];
+    entityValues.push(entityValue);
+
+    timeWindows.forEach((timeWindow) => {
+      const anomalySummaryInTimeRange = anomaliesSummary.filter(
+        (singleAnomalySummary) =>
+          singleAnomalySummary.startTime >= timeWindow.startDate &&
+          singleAnomalySummary.startTime < timeWindow.endDate
+      );
+
+      if (isEmpty(anomalySummaryInTimeRange)) {
+        maxAnomalyGradesForEntity.push(0);
+        numAnomalyGradesForEntity.push(0);
+        return;
+      }
+
+      const maxAnomalies = anomalySummaryInTimeRange.map((anomalySummary) => {
+        return toFixedNumberForAnomaly(
+          defaultTo(get(anomalySummary, 'maxAnomaly'), 0)
+        );
+      });
+      const countAnomalies = anomalySummaryInTimeRange.map((anomalySummary) => {
+        return defaultTo(get(anomalySummary, 'anomalyCount'), 0);
+      });
+
+      maxAnomalyGradesForEntity.push(Math.max(...maxAnomalies));
+      numAnomalyGradesForEntity.push(
+        countAnomalies.reduce((a, b) => {
+          return a + b;
+        })
+      );
+    });
+
+    maxAnomalyGrades.push(maxAnomalyGradesForEntity);
+    numAnomalyGrades.push(numAnomalyGradesForEntity);
+  });
+
+  const plotTimes = timeWindows.map((timeWindow) => timeWindow.startDate);
+  const timeStamps = plotTimes.map((timestamp) =>
+    moment(timestamp).format(HEATMAP_X_AXIS_DATE_FORMAT)
+  );
+  const plotData =
+    //@ts-ignore
+    {
+      x: timeStamps,
+      y: entityValues.reverse(),
+      z: maxAnomalyGrades.reverse(),
+      colorscale: ANOMALY_HEATMAP_COLORSCALE,
+      //@ts-ignore
+      zmin: 0,
+      zmax: 1,
+      type: 'heatmap',
+      showscale: false,
+      xgap: 2,
+      ygap: 2,
+      opacity: 1,
+      text: numAnomalyGrades.reverse(),
+      hovertemplate:
+        '<b>Time</b>: %{x}<br>' +
```

Comment:
Can we abstract these lines into const or function?

Reply:
sure. will do that.